src/PluginForm/Bluesnap/PaymentMethodAddForm.php

```diff
@@ -0,0 +1,176 @@
+    // The BlueSnap unique Hosted Payment Fields Token that will be sent in the
+    // drupalSettings to the JS.
+    // First, initialize BlueSnap.
+    if (empty($form_state->getValue('bluesnap_token'))) {
+      $bluesnap_token = $this->apiService->getHostedPaymentFieldsToken($this->entity->getPaymentGateway()->getPlugin());
+    }
+    else {
+      $bluesnap_token = $form_state->getValue('bluesnap_token');
+    }
+
+    // Alter the form with Bluesnap specific needs.
+    $element['#attributes']['class'][] = 'bluesnap-form';
+    $element['#attached']['library'][] = 'commerce_bluesnap/form';
+    $element['#attached']['drupalSettings']['bluesnap_token'] = $bluesnap_token;
+
+    // Hidden fields which will be populated by the js.
+    $element['bluesnap_token'] = [
+      '#type' => 'hidden',
+      '#attributes' => [
+        'id' => 'bluesnap-token',
+      ],
+    ];
+    // Hidden fields which will be populated by the js once a card is
+    // successfully submitted to BlueSnap. This will be used for the card
+    // details for Anonymous users when creating a payment method.
+    $element['bluesnap_cc_type'] = [
+      '#type' => 'hidden',
+      '#attributes' => [
+        'id' => 'bluesnap-cc-type',
+      ],
+    ];
+    $element['bluesnap_cc_last_4'] = [
+      '#type' => 'hidden',
+      '#attributes' => [
+        'id' => 'bluesnap-cc-last-4',
+      ],
+    ];
+    $element['bluesnap_cc_expiry'] = [
+      '#type' => 'hidden',
+      '#attributes' => [
+        'id' => 'bluesnap-cc-expiry',
+      ],
+    ];
+
+    // The credit card fields necessary for BlueSnap.
+    $element['card_number'] = [
+      '#type' => 'item',
+      '#title' => t('Card number'),
+      '#required' => TRUE,
+      '#validated' => TRUE,
+      '#markup' => '<div class="form-control" id="card-number" data-bluesnap="ccn"></div>',
+    ];
+
+    $element['expiration'] = [
+      '#type' => 'item',
+      '#title' => t('Expiration date'),
+      '#required' => TRUE,
+      '#validated' => TRUE,
+      '#markup' => '<div class="form-control" id="exp-date" data-bluesnap="exp"></div>',
+    ];
+
+    $element['security_code'] = [
+      '#type' => 'item',
+      '#title' => t('CVV'),
+      '#required' => TRUE,
+      '#validated' => TRUE,
+      '#markup' => '<div class="form-control" id="cvv" data-bluesnap="cvv"></div>',
+    ];
+
+    // To display validation errors.
+    $element['payment_errors'] = [
+      '#type' => 'markup',
+      '#markup' => '<div id="payment-errors"></div>',
+      '#weight' => -200,
+    ];
+
+    return $element;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function validateCreditCardForm(array &$element, FormStateInterface $form_state) {
+    // The JS library performs its own validation.
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitCreditCardForm(array $element, FormStateInterface $form_state) {
+    // The payment gateway plugin will process the submitted payment details.
+  }
```

Comment:
**CHANGE REQUEST**

*Readability/Coding standards*

Sorry, line lengths even here, including comments and function arguments on both functions above.